### PR TITLE
Common: add apml service and command

### PR DESCRIPTION
**Pull request by Yi-Shum**
Summary:
- Support APML.
- Add OEM commands to access APML.(Netfn 0x38, command 0x2C-0x2F)

Test Plan:
Build code: PASS
Command test: PASS

Log:
```
write/read RMI:
root@bmc-oob:~# bic-util slot1 0xe0 0x2c 0x15 0xa0 0x00 0x0 0x38
15 A0 00 00
root@bmc-oob:~# bic-util slot1 0xe0 0x2d 0x15 0xa0 0x00 0x0 0x38 0x01
15 A0 00
root@bmc-oob:~# bic-util slot1 0xe0 0x2c 0x15 0xa0 0x00 0x0 0x38
15 A0 00 01

write/read TSI:
root@bmc-oob:~# bic-util slot1 0xe0 0x2c 0x15 0xa0 0x00 0x01 0x07
15 A0 00 46
root@bmc-oob:~# bic-util slot1 0xe0 0x2d 0x15 0xa0 0x00 0x01 0x07 0x57
15 A0 00
root@bmc-oob:~# bic-util slot1 0xe0 0x2c 0x15 0xa0 0x00 0x01 0x07
15 A0 00 57

access Mailbox:
root@bmc-oob:~# bic-util slot1 0xe0 0x2e 0x15 0xa0 0x00 0x00 0x01 0x00 0x00 0x00 0x00
15 A0 00 00
root@bmc-oob:~# bic-util slot1 0xe0 0x2f 0x15 0xa0 0x00 0x00
15 A0 00 01 0D F8 01 00 00

access CPUID:
root@bmc-oob:~# bic-util slot1 0xe0 0x2e 0x15 0xa0 0x00 0x01 0x00 0x00 0x00 0x00 0x00 0x00
15 A0 00 01
root@bmc-oob:~# bic-util slot1 0xe0 0x2f 0x15 0xa0 0x00 0x01
15 A0 00 00 10 00 00 00 41 75 74 68

access MCA:
root@bmc-oob:~# bic-util slot1 0xe0 0x2e 0x15 0xa0 0x00 0x02 0x00 0x00 0x00 0x00 0x00
15 A0 00 02
root@bmc-oob:~# bic-util slot1 0xe0 0x2f 0x15 0xa0 0x00 0x02
15 A0 00 45 10 00 00 00 41 75 74 68
```

### CODE REVIEW

File: common/service/apml/apml.c [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+		k_msleep(WAIT_TIME_MS);
+	}
+	return false;
+}
+/****************** MCA *********************/
+
+static uint8_t write_MCA_request(apml_msg *msg)
+{
+	uint8_t retry = 5;
+	I2C_MSG i2c_msg;
```
Comment (GoldenBug) [inline common/service/apml/apml.c:71]: Verify msg is not NULL when passed in.

File: common/service/apml/apml.c [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+	if (i2c_master_write(&i2c_msg, retry)) {
+		return APML_ERROR;
+	}
+	return APML_SUCCESS;
+}
+
+static uint8_t read_MCA_response(apml_msg *msg)
+{
+	uint8_t retry = 5;
+	I2C_MSG i2c_msg;
```
Comment (GoldenBug) [inline common/service/apml/apml.c:90]: Verify msg is not NULL when passed in.

File: common/service/apml/apml.c [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+	if (i2c_master_read(&i2c_msg, retry)) {
+		return APML_ERROR;
+	}
+
+	memcpy(msg->RdData, &i2c_msg.data[1], sizeof(mca_RdData));
+	return APML_SUCCESS;
+}
+
+static uint8_t access_MCA(apml_msg *msg)
+{
```
Comment (GoldenBug) [inline common/service/apml/apml.c:105]: Verify msg is not NULL

File: common/service/apml/apml.c [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+	uint8_t retry = 5;
+	I2C_MSG i2c_msg;
+	i2c_msg.bus = msg->bus;
+	i2c_msg.target_addr = msg->target_addr;
+	i2c_msg.tx_len = 10;
+	i2c_msg.data[0] = 0x73;
+	i2c_msg.data[1] = 0x08;
+	i2c_msg.data[2] = 0x08;
+	i2c_msg.data[3] = 0x91;
+	memcpy(&i2c_msg.data[4], msg->WrData, sizeof(cpuid_WrData));
```
Comment (GoldenBug) [inline common/service/apml/apml.c:140]: verify msg is not NULL when passed into function

File: common/service/apml/apml.c [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+
+	if (i2c_master_write(&i2c_msg, retry)) {
+		return APML_ERROR;
+	}
+	return APML_SUCCESS;
+}
+
+static uint8_t read_CPUID_response(apml_msg *msg)
+{
+	uint8_t retry = 5;
```
Comment (GoldenBug) [inline common/service/apml/apml.c:150]: Same, verify msg not NULL

File: common/service/apml/apml.c [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+			}
+			break;
+		default:
+			break;
+		}
+
+		if (msg_cfg.cb_fn) {
+			msg_cfg.cb_fn(&msg_cfg);
+		}
+		k_msleep(10);
```
Comment (GoldenBug) [inline common/service/apml/apml.c:376]: It looks like this code is only executed when none of the other msg types are executed.
If that is true we should put it in the `default` case in the switch statement.

**Review by GoldenBug — CHANGES_REQUESTED** [review]
A lot of functions aren't checking the input pointer parameters to verify they are not NULL. Please add these checks and address the other comments.

File: common/service/apml/apml.c (hunk 1 above)
Comment (Yi-Shum) [reply common/service/apml/apml.c:71]: Will check if msg is NULL.

File: common/service/apml/apml.c (hunk 2 above)
Comment (Yi-Shum) [reply common/service/apml/apml.c:90]: Will check if msg is NULL.

File: common/service/apml/apml.c (hunk 3 above)
Comment (Yi-Shum) [reply common/service/apml/apml.c:105]: Will check if msg is NULL.

File: common/service/apml/apml.c (hunk 4 above)
Comment (Yi-Shum) [reply common/service/apml/apml.c:140]: Will check if msg is NULL.

File: common/service/apml/apml.c (hunk 5 above)
Comment (Yi-Shum) [reply common/service/apml/apml.c:150]: Will check if msg is NULL.

File: common/service/apml/apml.c (hunk 6 above)
Comment (Yi-Shum) [reply common/service/apml/apml.c:376]: It should be executed after accessing every msg_type except default.
I will modify code to achieve this.

**Comment by Yi-Shum** [comment]
Hi @GoldenBug ,
I modified it based on the comments.
Please review again, thank you

**Pull request closed without merging**